release-calendar/src/server/repository-service.ts

```diff
@@ -57,4 +57,58 @@ export class RepositoryService {
 
     return releaseQuery;
   }
+
+  include(arr: Release[], rel: Release): boolean {
+    for (let i = 0; i < arr.length; i++) {
+      if (arr[i].name === rel.name) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  async getCurrentReleases(): Promise<Release[]> {
+    const rollbackQuery = this.releaseRepository
+      .createQueryBuilder('release')
+      .innerJoin(
+        'release.promotions',
+        'promotion',
+        'promotion.channel = :channel',
+        {channel: 'rollback'},
+      )
+      .select('release.name')
+      .addSelect('promotion.channel')
+      .getMany();
+
+    const channelQueries = [
+      Channel.LTS,
+      Channel.NIGHTLY,
+      Channel.OPT_IN_BETA,
+      Channel.OPT_IN_EXPERIMENTAL,
+      Channel.PERCENT_BETA,
+      Channel.PERCENT_EXPERIMENTAL,
+      Channel.STABLE,
+    ].map((eachChannel) =>
+      this.releaseRepository
+        .createQueryBuilder('release')
+        .innerJoinAndSelect(
+          'release.promotions',
+          'promotion',
+          'promotion.channel = :channel',
+          {channel: eachChannel},
+        )
+        .select('release.name')
+        .addSelect('promotion.channel')
+        .orderBy('promotion.date', 'DESC')
+        .getMany(),
+    );
+
+    const results: Release[] = [];
+    const rollback = await rollbackQuery;
+    const channels = await Promise.all(channelQueries);
+    channels.forEach((channel) =>
+      results.push(channel.find((first) => !this.include(rollback, first))),
```

Comment:
How is `channel.find()` working? Isn't `channel` in this case, type `Release`?

Reply:
`channel.find()` is scanning the releases in that channel that have never also been rolled back. It's doing a scan from the first to the last, which because I queried the releases in each channel to be in chronologically descending order will start at the most recent. Because of rollbacks happen infrequently, find should only be checking the first few releases that have came out in each channel.

Reply:
I did go back and try to clarify the variables. 
```const currentReleases = releasesInEachChannel.map((releasesInOneChannel) =>
      releasesInOneChannel.find(
        (latestRelease) => rollback.indexOf(latestRelease) == -1,
      ),
    );'''

Reply:
Nice! thanks